### DISCUSSION TITLE
how to calculate the transaction hash use Typescript SDK?

### DISCUSSION THREAD

**Discussion by glant509** [aptos-labs/aptos-developer-discussions#104] [Questions] [answered]
how to calculate the transaction hash use Typescript SDK?

**Comment by kshitijc1506** [comment]
To get the transaction hash using the TypeScript SDK, you would follow the process of 

1. Creating a transaction
2. Signing it
3. Submitting it to the Aptos Blockchain (you can do step 2 and 3 together with `aptos.signAndSubmitTransaction`)

Once the transaction is submitted, the API will return a transaction hash that represents the unique identifier of the transaction on-chain. Here's an example of what that looks like:

```
{
  hash: '0x2a8e0ec1cff74c7021b09259632896f0f2636b1b952b2d595a3e831e4ee79fd6',
  sender: '0xe4078b6fc336b4c0fb53ecc3458f5a9df97b99c433ada72334f952f7383716a8',
  sequence_number: '0',
  max_gas_amount: '200000',
  gas_unit_price: '100',
  expiration_timestamp_secs: '1711462618',
  payload: {
    function: '0x1::aptos_account::transfer_coins',
    type_arguments: [ '0x1::aptos_coin::AptosCoin' ],
    arguments: [
      '0xce54cdc54d837abeb019aa3f97a824b0a60fb7443fe4ca48bd6c0086911003a6',
      '0x6400000000000000'
    ],
    type: 'entry_function_payload'
  },
  signature: {
    public_key: '0xf7f983d47664c226dfed875f3238fcc4a2d43848ca6133015ff43f55d512c3d3',
    signature: '0xdcf82e619dd01fbc943b62fe721dd37e7d7e6ad3088a7ff48cc4dec23c768ff9473865b3dbc69e69fcdf7c7c3d752487cb3441b69295ad14688e97019010b00b',
    type: 'ed25519_signature'
  }
}
```

Refer to these [docs](https://aptos.dev/sdks/ts-sdk/#build-and-submit-transaction) for more info.

**Reply by glant509** [reply]
thank you, I boradcast the signed transaction to the blockchain use a http request, like this: curl --location 'https://api.mainnet.aptoslabs.com/v1/transactions' \
--header 'Accept: application/json, application/x-bcs' \
--header 'Content-Type: application/json' \
--header 'Cookie: GCLB=CLSjkdzEytuBSxAD' \
--data '{"sender":"0xddefc66c31eb4f905e32f1597ced16d9906c21d32ec2b560c705025a390b6d9e","sequence_number":"9","max_gas_amount":"200000","gas_unit_price":"100","expiration_timestamp_secs":"1711594676","payload":{"function":"0x1::aptos_account::transfer_coins","type_arguments":["0x1::aptos_coin::AptosCoin"],"arguments":["0x829bd5dd49a5b139d5fd559c8b2db9cf496e02998161a3b68ba6dec4c87e94fd","121"],"type":"entry_function_payload"},"signature":{"type":"ed25519_signature","public_key":"0xc7fde1d1a497936a4ba80c31486d52372f00bc9742210c2b0ecb8920661b595d","signature":"0x2dc6a6bf0a17c5487268808df278b6b185d64955c4cb8494c07b0213e6ddf520390137fbc2d8e8f20b5ece059de97be332b8e0fca7e47bdb93491b15b9b53c05"}}' , when I use a http request to broadcast the signed transaction, the response coce is 202, and the response body is empty, I cann;t get the transaction hash directly, so I need to calculate the transaction hash. I cann't used the typescript sdk to broadcast the signed transaction because the sign server is in a inner system, it unable to access external network.so I need to calculate the transaction hash by myself or use the Typescript sdk,  this is my code to sign a transaction, how to calculate the transaction hash with the sdk without use the broadcast api?  here is my code :  let simpleTransaction = await aptos.transferCoinTransaction({
        sender: ed25519Account.accountAddress,
        recipient: toAddress,
        amount: amount,
        coinType: "0x1::aptos_coin::AptosCoin",
        options: {
            maxGasAmount: max_gas_amount,
            gasUnitPrice: gas_unit_price,
            expireTimestamp: expiration_timestamp_secs,
            accountSequenceNumber: sequence_number,
        }
    });
    const senderAuthenticator = aptos.transaction.sign({
        signer: ed25519Account,
        transaction: simpleTransaction,
    });


    console.log("publicKey: ", ed25519Account.publicKey.toString());
    console.log("signature: ", senderAuthenticator.signature.data.toString())
    console.log("expire", expiration_timestamp_secs)
    console.log("hexdata: ", senderAuthenticator.bcsToHex().toString());

**Comment by gregnazario** [comment, marked as answer]
We've got a new answer for this one!  I've copied this in a stack overflow as well.

## Example

https://github.com/aptos-labs/aptos-ts-sdk/blob/5cec8cecb6684ec37d379a62ae8df6a6fab06f61/tests/e2e/transaction/transactionBuilder.test.ts#L559-L579

## Short answer (how to hash)

This has now been added to the TypeScript SDK, take a look here
https://github.com/aptos-labs/aptos-ts-sdk/commit/5cec8cecb6684ec37d379a62ae8df6a6fab06f61

It basically does:

```
let prefix = SHA3-256("APTOS::Transaction")

SHA3-256([prefix, 0 byte, BCS encoded signed transaction])
```

## Longer answer (How to hash)

Aptos hashes all types in a standard way.  This is done by first creating a prefix, for every type.  This is defined as a SHA3-256 of `APTOS::<Type>`
The struct type is here as `Transaction` https://github.com/aptos-labs/aptos-core/blob/main/types/src/transaction/mod.rs#L1931-L1959
```
let prefix = SHA3-256("APTOS::Transaction")
```

Then, there is serialization of the transactions as BCS.  In the case here, the TypeScript SDK drops the enum byte, so it has to be recombined (in BCS).  Take a look here at the details https://github.com/aptos-labs/aptos-core/blob/main/types/src/transaction/mod.rs#L1931-L1959

The order of the items in the struct give the enum byte.

```
let enumByte = 0; // UserTransaction
let userTransaction = [enumByte, transactionBcsEncoded];
```

Finally, hashing all of these together gives the final answer
```
let transactionHash = SHA3-256(prefix + userTransaction)
```
